File: app/src/main/java/com/pr0gramm/app/api/pr0gramm/HttpErrorException.java
package com.pr0gramm.app.api.pr0gramm;

import java.io.IOException;

import retrofit2.HttpException;

/**
 */
public class HttpErrorException extends IOException {
    private final String body;

    HttpErrorException(HttpException cause, String body) {
        super(cause);
        this.body = body;
    }

    public String getErrorBody() {
        return body;
    }

    @Override
    public HttpException getCause() {
        return (HttpException) super.getCause();
    }

    public int code() {
        return getCause().code();
    }

    /**
     * Creates a new exception by reading the response of the given one.
     */
    public static HttpErrorException from(HttpException cause) {
        String body = "error body not available";
        try {
            body = cause.response().errorBody().string();
        } catch (Exception ignored) {
        }

        return new HttpErrorException(cause, body);
    }
}

File: app/src/main/java/com/pr0gramm/app/api/pr0gramm/HttpErrorException.kt
package com.pr0gramm.app.api.pr0gramm

import retrofit2.HttpException
import java.io.IOException

/**
 */
class HttpErrorException(override val cause: HttpException, val errorBody: String) : IOException(cause) {
    val code = cause.code()

    companion object {
        /**
         * Creates a new exception by reading the response of the given one.
         */
        fun from(cause: HttpException): HttpErrorException {
            var body = "error body not available"
            try {
                body = cause.response().errorBody().string()
            } catch (ignored: Exception) {
            }

            return HttpErrorException(cause, body)
        }
    }
}
